app/main/api/handlers/generateAddress.ts

```diff
@@ -0,0 +1,152 @@
+  system: System, params: any
+): Promise<JsonSerializable> {
+  return Promise.resolve(params)
+    .then(parseParams)
+    .then(inject(validateParams, system))
+    .then(inject(createFinalKeyAddress, system))
+    .then(resultAsResponse)
+    .catch(buildGenerateAddressError)
+    .then((response) => asObject(response, GenerateAddressResponse))
+}
+
+/** Parse params into the GenerateAddressParams type */
+async function parseParams(params: any): Promise<GenerateAddressParams> {
+  try {
+    return asRuntimeType(params, GenerateAddressParams)
+  } catch (e) {
+    throw generateAddressErrors.WRONG_PARAMS
+  }
+}
+
+/** Validate that the parsed parsed params are semantically correct */
+async function validateParams(
+  params: GenerateAddressParams,
+  system: System
+): Promise<[GenerateAddressParams, Wallet]> {
+  const { account, requestedAmount, expirationDate } = params
+  const wallet = system.appStateManager.state.wallet
+  const now = Math.floor(Date.now() / 1000)
+
+  if (requestedAmount && requestedAmount < 0) {
+    throw generateAddressErrors.NEGATIVE_AMOUNT
+  }
+
+  if (expirationDate) {
+    if (expirationDate < now) {
+      throw generateAddressErrors.PAST_EXPIRATION_DATE
+    }
+    if (expirationDate > now + MAX_EXPIRATION_DATE_DIFF) {
+      throw generateAddressErrors.TOO_FAR_EXPIRATION_DATE
+    }
+  }
+
+  if (!wallet) {
+    throw generateAddressErrors.NO_UNLOCKED_WALLET
+  }
+
+  if (!wallet.accounts[account]) {
+    throw generateAddressErrors.WRONG_ACCOUNT
+  }
+
+  return [params, wallet]
+}
+
+/** Generate a final key, store it, and return its corresponding address  */
+async function createFinalKeyAddress(
+  [params, wallet]: [GenerateAddressParams, Wallet],
+  system: System
+): Promise<ExternalFinalKey> {
+  try {
+    const { account, ...metadata } = { ...params }
+
+    const creationDate = Math.floor(Date.now() / 1000)
+    // const account = params.account
+    const seed = wallet.seed.seed
+    const masterKey: key.ExtendedKey<privateKey.PrivateKey> = {
+      chainCode: seed.chainCode,
+      key: {
+        type: "private",
+        bytes: seed.masterSecret
+      }
+    }
+    const finalKeys = wallet.accounts[account].keyChains[0].finalKeys
+    const finalKeyIndex = finalKeys.length
+    const path = `m/3'/4919'/${account}'/0/${finalKeyIndex}`
+    const extendedFinalPrivateKey = privateKey.derive(masterKey, path)
+    const keyPathStruct = keyPath.fromString(path)
+    const extendedFinalPublicKey = publicKey.create(extendedFinalPrivateKey)
+    const address = p2pkh.encode(extendedFinalPublicKey.key, ChainType.test)
+    const pkh = p2pkh.decode(address)[1]
+    const externalFinalPrivKey: ExternalFinalKey = {
+      kind: "external",
+      extendedKey: {
+        type: "private",
+        chainCode: extendedFinalPrivateKey.chainCode,
+        key: extendedFinalPrivateKey.key.bytes
+      },
+      keyPath: keyPathStruct,
+      pkh,
+      metadata: {
+        ...metadata,
```

Comment:
the [issue description](https://github.com/witnet/sheikah/issues/391) says something about setting the label when it comes as undefined

Reply:
Good one!